### PR TITLE
feat: ci: automate the new release process

### LINKED ISSUE #12162: [DX Streamline] Use the mkreleaselog generated changelog when creating GitHub Releases
## Problem to solve
When a GitHub Release of lotus is created, it contains an auto-generated changelog. Then, a release performer has to update the release description manually to reflect the mkrelease log outputs (and CHANGELOG.md) contents. This should be automated.

## Why Important
There is no good reason to put the demand of running a script and copying its' outputs on a human. Computers are capable of doing that.

## User/Customer
Lotus maintainers

## Notes
Enabled by https://github.com/filecoin-project/lotus/pull/12096

### LINKED ISSUE #12163: [DX Streamline] Create draft GitHub releases on PRs modifying version.go
## Problem to solve
Right now, a GitHub Release is created only after a PR modifying version.go is merged. This makes it impossible to verify the release or update its description prior to the merge. It also means that the release, on publish, is not yet ready for general public fully (lacking proper description).

## Why Important
As soon as a really is out, it should be final - including the description part. By creating drafts on PRs, we'll be able to prepare all the aspects of the release prior to merge.

## User/Customer
Lotus maintainers

## Notes
Enabled by https://github.com/filecoin-project/lotus/pull/12096

### PR DESCRIPTION
**Pull request by galargh**
Closes #12163
Closes #12162

## Related Issues

https://github.com/filecoin-project/lotus/issues/12010

## Proposed Changes

0. Split node and miner releases (node=[lotus], miner=[lotus-miner, lotus-worker]).
1. Publish a new GitHub release (and create an appropriate tag) of node and/or miner whenever a PR which modified `version.go` (in a way that adds a version that doesn't have a corresponding tag yet) has been merged to either releases or release/vX.Y.Z branch.
2. Run the release workflow in a dry/snapshot mode on PRs that modify `version.go` (if -rc, then base has to be release/vX.Y.Z, if stable ,then base has to be releases).
3. Run the docker workflow in a dry mode on PRs. 

❗ I split `homebrew-lotus` into 2 formulas (for lotus node - `lotus` and lotus miner - `lotus-miner`).

## Additional Info

Out of scope for this PR, but worth considering in the future:
1. https://github.com/filecoin-project/lotus/issues/12162
2. https://github.com/filecoin-project/lotus/issues/12163

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] If the PR affects users (e.g., new feature, bug fix, system requirements change), update the CHANGELOG.md and add details to the UNRELEASED section.
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: cmd/release/main.go
Comment (rvagg) [inline cmd/release/main.go:1]: this is a nicer idea than my source grepping proposal
just make sure this is well documented so it's clear what this is doing in here and that it has a very narrow usecase

**Comment by BigLep** [comment]
I think it would be good if this PR also included the accompanying changes to https://github.com/filecoin-project/lotus/blob/master/documentation/misc/RELEASE_ISSUE_TEMPLATE.md so its clear how this affects the "release engineer".  Maybe @rjan90 should take this?

File: cmd/release/main.go
Comment (galargh) [reply cmd/release/main.go:1]: On it, I'll add a README to that package and document the interface and the purpose of the CLI.

**Comment by galargh** [comment]
> I think it would be good if this PR also included the accompanying changes to https://github.com/filecoin-project/lotus/blob/master/documentation/misc/RELEASE_ISSUE_TEMPLATE.md so its clear how this affects the "release engineer". Maybe @rjan90 should take this?

That's a good idea, I'll propose the changes to that file - I think it'll be good way to do the final verification on our mutual understanding of the new process.

**Comment by galargh** [comment]
I updated the documentation and the accompanying workflows to make sure the automation changes are reflected.

File: .goreleaser.yaml
Comment (BigLep) [inline .goreleaser.yaml:86]: Are these p0 todos items that need to be handled before merge?  If so, maybe pull out as PR checklist items or self-comment so it's clear there are unresolved comments?

File: cmd/release/README.md
Comment (BigLep) [inline cmd/release/README.md:3]:
```suggestion
The Lotus Release Tool is a CLI (Command Line Interface) utility designed to facilitate interactions with the Lotus Node and Miner metadata. This tool allows users to retrieve version information in either JSON or text format. The Lotus Release Tool was developed as a part of the [2024Q2 release process automation initiative](https://github.com/filecoin-project/lotus/issues/12010) and is used in the GitHub Actions workflows related to the release process.
```

File: cmd/release/README.md
Comment (BigLep) [inline cmd/release/README.md:9]: What logs?

**Review by BigLep — COMMENTED** [review]
It looks good to me from my high level read through.  We'll want @rvagg to give the approval here though.

In general, and this may just be me, I would like to see less bash for future maintenance regards.  That said, I know there are lots of shell scripts in https://github.com/filecoin-project/lotus/tree/master/scripts and I assume that is the convention within github workflow .yml files.  Anyways, no insistence on changing this - I just wanted to make sure that it's intentional when we add more shell scripting given it can tend to be more esoteric and less ergonomic than other languages.

File: cmd/release/README.md
Comment (BigLep) [reply cmd/release/README.md:9]: Oh, I misread this.  It means the output can be printed in JSON or text.  Got it.

File: .github/workflows/docker.yml
Comment (rvagg) [inline .github/workflows/docker.yml:27]: what's the implication of this change? do we get docker builds for all branches now?

File: .github/workflows/release.yml
Comment (rvagg) [inline .github/workflows/release.yml:107]:
```suggestion
      - if: matrix.project == 'node' && runner.os == 'macOS'
```

File: .github/workflows/release.yml
Comment (rvagg) [inline .github/workflows/release.yml:59]:
```suggestion
              projects+=($project)
              echo "Adding $project to build list"
```

Not sure how helpful this is ultimately but as long as we're debugging skip cases

File: .github/workflows/release.yml
Comment (rvagg) [inline .github/workflows/release.yml:160]:
```suggestion
          if [[ "${{ matrix.project}}" == "node" ]]; then
            tag="$version"
            latest="$(git tag -l 'v*' | sort -V | tail -n1)"
          elif [[ "${{ matrix.project}}" == "miner" ]]; then
            tag="miner/$version"
            latest="$(git tag -l 'miner/v*' | sort -V | tail -n1)"
```

mind if we switch these to remain consistent in ordering?

File: .github/workflows/release.yml
Comment (rvagg) [inline .github/workflows/release.yml:177]: nice, this is new, right? have we tested this?

**Comment by rvagg** [comment]
👍 so far

File: .github/workflows/release.yml
Comment (galargh) [reply .github/workflows/release.yml]: Yes, now we did - I forgot the binaries are for different archs. I now moved the tests back to the build jobs.

File: .github/workflows/release.yml
Comment (galargh) [reply .github/workflows/release.yml]: I removed that part completely after testing.

File: .github/workflows/docker.yml
Comment (galargh) [reply .github/workflows/docker.yml:27]: What's new is building docker images on pull requests to master and release branches (they are not published anywhere though). All the publishing remains as it was before.

File: .goreleaser.yaml
Comment (galargh) [reply .goreleaser.yaml:87]: Yes, that was exactly the idea. I think I've resolved all of them now. Apart from this one, but that's not a p0 because we don't create drafts at all in this iteration. I'll update that in a sec.

**Comment by galargh** [comment]
> In general, and this may just be me, I would like to see less bash for future maintenance regards.

I "debashified" it a little bit ;) 

I've also performed testing both in a fork and in this repo (I had access to a proper goreleaser license here). The only areas not fully covered by tests are homebrew-lotus updates and publishing non-draft releases. The former will be easy to fix manually if it doesn't go as expected - which is why I think we can leave the testing to when we do a proper release. The latter is no different conceptually from releasing a draft, so it should be fine, too. 

Of course, I'll be available whenever we start doing the releases with the updated workflows. 

All in all, ready for review :)

**Comment by BigLep** [comment]
Thanks @galargh.
I assume we'll try using this for 1.27.2: https://github.com/filecoin-project/lotus/issues/12157

**Comment by galargh** [comment]
> Thanks @galargh. @rvagg: I assume we'll try using this for 1.27.2: #12157

FYI, if we do, we have to first cherry-pick this PR onto the target branches so that when we make the accompanying PRs, the version of the workflow that runs is the one that we updated here.

**Comment by BigLep** [comment]
2024-07-03 notes form conversation with IPDX: I propose we merge this into master.  Per above, it can be cherry-picked for releases like 1.27.2.  (@galargh is happy to help with that if desired.). I'd like to use it with 1.27.2 so we can work out any kinks while this this is fresh.
Does that sound good @rvagg?  If so, can you please merge.

**Comment by rvagg** [comment]
https://github.com/filecoin-project/lotus/pull/12194

File: .goreleaser.yaml
Comment (rvagg) [inline .goreleaser.yaml:86]: should we set this to `true` so we don't end up with multiple drafts if re-pushing the same branch?

**Comment by rvagg** [comment]
* I merged a version of this, https://github.com/filecoin-project/lotus/pull/12194 into `release/v1.27.2`
* Updated my PR into that branch with the content to be `v1.27.2-rc1` https://github.com/filecoin-project/lotus/pull/12159
* The release workflow has been triggered, artifacts built, goreleaser run, but it's not done anything useful with it: https://github.com/filecoin-project/lotus/actions/runs/9867424047/job/27247851095?pr=12159

Is this what we're expecting? I see that we're not setting `GORELEASER_KEY` or `GITHUB_TOKEN` if `PUBLISH` is not `true`, so it seems that this is intended behaviour. So is the `draft: false` in `.goreleaser.yml` just a protection mechanism, and we're not intending it to publish a draft release to GH?

I'll merge this and 🤞 that this produces a release.

**Comment by rvagg** [comment]
Merged, it ran, built and tried to publish: https://github.com/filecoin-project/lotus/actions/runs/9867547199/job/27248227319

We got a draft release with 2 of the assets: https://github.com/filecoin-project/lotus/releases/tag/untagged-04b7f2696e2852f8df23

But a failure in in the action: https://github.com/filecoin-project/lotus/actions/runs/9867547199/job/27248227319

```
  ⨯ release failed after 37s                 error=scm releases: failed to publish artifacts: could not update existing release: PATCH https://api.github.com/repos/filecoin-project/lotus/releases/164711034: 422 Validation Failed [{Resource:Release Field:pre_receive Code:custom Message:pre_receive Repository rule violations found

Cannot create ref due to create name restrictions.

} {Resource:Release Field: Code:custom Message:Published releases must have a valid tag}]
```

1. Why did we not get the arm64 artifact?
2. I guess we should build tagging into this process too?

**Comment by rvagg** [comment]
Also interesting that it's just a draft, does goreleaser make a draft first, populate it with artifacts, and then publish it? Or is this intended behaviour? The current release text is just a raw changelog, which isn't great and it'll need editing, so maybe draft is good and we can manually publish after cleaning it up. Alternatively, capture the changelog content for the release text since that's what's going to be used anyway.

**Comment by galargh** [comment]
> Why did we not get the arm64 artifact?

We always get only 2 assets. One for linux x64, and one universal binary for macOS which is intended to work both on x64 and arm64.

> I guess we should build tagging into this process too?

It's already built in. The tagging happens automatically when the GitHub release is published. That's actually the problem we're facing now - tags are restricted so the publish of the release gets blocked and causes the workflow to fail. We could fix this by lifting the tag restriction, but I also noticed a few other minor issues in the run so I'm going to address them first.

> Also interesting that it's just a draft, does goreleaser make a draft first, populate it with artifacts, and then publish it?

Yes, exactly. Goreleaser first creates a draft and then publishes it and that's when the failure happens.

> Alternatively, capture the changelog content for the release text since that's what's going to be used anyway.

That's exactly what I want to do. With the new process of keeping changelog up to date, we can totally automate the release notes. I also want to include the mkreleaselog outputs in there, too. 

---

@rvagg 🙇 Thank you for testing this, it's super helpful and I see where the shortcomings of the process are. I already have the improvements ready, I "just" need to test them. I'll try to be done with it today (tomorrow the latest). I'll merge them into this branch when ready. I'll ping you directly when that happens.

**Comment by rvagg** [comment]
> We always get only 2 assets. One for linux x64, and one universal binary for macOS which is intended to work both on x64 and arm64.

oh, you're right! I didn't even notice we didn't do an arm64 linux; I guess nobody's asking for it, but that seems a little odd.

**Comment by galargh** [comment]
Done! Thanks to removing the dependency on goreleaser I was able to fully test the release flow in a fork, which gives me confidence that this is now ready for merge.

New features:
- on PRs to `release/v*`/`release/miner/v*` draft GitHub releases for RC releases will be created
- on PRs to `releases` draft GitHub releases for non-RC releases will be created
- on pushes to `release/v*`/`release/miner/v*` GitHub releases for RC releases and their corresponding tags will be created
- on pushes to `releases` GitHub releases for non-RC releases and their corresponding tags will be created
- the created releases will contain pre-populated notes consisting of their entry from the CHANGELOG.md (or the unreleased section) and the output of mkreleaselog

Question:
1. This iteration removes updating https://github.com/filecoin-project/homebrew-lotus on lotus releases. As per https://formulae.brew.sh/api/analytics/install/365d.json lotus was installed 184 times in the last 365 days via homebrew (11 in last 30 days). Would you like us to automate the update of that repo?

**Comment by BigLep** [comment]
@galargh : I'll let others chime in here, but a couple of comments or things to be aware of.

1. FYI that we'll be dropping the "releases" branch soon (after the nv23 upgrade) as part of https://github.com/filecoin-project/lotus/issues/12072

2. I believe we should maintain the Homebrew functionality given it is the recommended path on Macs: https://lotus.filecoin.io/lotus/install/macos/

**Comment by rvagg** [comment]
https://github.com/filecoin-project/lotus/pull/12231

File: documentation/misc/RELEASE_ISSUE_TEMPLATE.md
Comment (rjan90) [inline documentation/misc/RELEASE_ISSUE_TEMPLATE.md:22]:
```suggestion
Perform the following changes to the `release/vX.Y.Z` or `release/miner/vX.Y.Z` branch through a PR:
- [ ] Update the version string in `build/version.go` to one ending with '-rcX'. Ensure you update the appropriate version string based on whether you are creating a node release (`NodeBuildVersion`), a miner release (`MinerBuildVersion`) or both.
```

Comment (rvagg) [inline documentation/misc/RELEASE_ISSUE_TEMPLATE.md:22]:
```suggestion
- [ ] Update the version string in `build/version.go` to one ending with '-rcX'. Ensure you update the appropriate version string based on whether you are creating a node release (`NodeBuildVersion`), a miner release (`MinerBuildVersion`) or both.
```

from @rjan90 in https://github.com/filecoin-project/lotus/pull/12231#discussion_r1677505512

**Comment by galargh** [comment]
> 1. FYI that we'll be dropping the "releases" branch soon (after the nv23 upgrade) as part of [Revised Lotus Client Versioning #12072](https://github.com/filecoin-project/lotus/issues/12072)

Off the top of my head, getting rid of `releases` will require:
- updating triggers - e.g. https://github.com/filecoin-project/lotus/blob/1b9f8126b1de50472a0701f6bffffc771e78ef1d/.github/workflows/release.yml#L8
- simplifying this check - https://github.com/filecoin-project/lotus/blob/1b9f8126b1de50472a0701f6bffffc771e78ef1d/.github/workflows/release.yml#L61-62

And that should be about it.

> 2. I believe we should maintain the Homebrew functionality given it is the recommended path on Macs: https://lotus.filecoin.io/lotus/install/macos/

OK, we'll set it up this will in the homebrew-lotus repository directly.

---

I don't think these two are blocking.

**Comment by BigLep** [comment]
I assume this is waiting on final approval from @rvagg .  I've re-requested review.

**Review by rvagg — APPROVED** [review]
👌 this worked well in https://github.com/filecoin-project/lotus/pull/12232 to produce https://github.com/filecoin-project/lotus/releases/tag/v1.27.2-rc2

I'm happy with this if @rjan90 is.

**Comment by rvagg** [comment]
Some issues encountered when taking v1.27.2 all the way that will need to be addressed here: https://github.com/filecoin-project/lotus/pull/12249#issuecomment-2232977928

**Comment by galargh** [comment]
> This required a further merge of release/v1.27.2 into the releases branch to finalise https://github.com/filecoin-project/lotus/pull/12250, which is different from RC builds. This is as designed but needs to change for v1.28.0 forward because we are not using the releases branch for that.

This was expected as per https://github.com/filecoin-project/lotus/pull/12096#issuecomment-2228268449. Updated now in https://github.com/filecoin-project/lotus/pull/12096/commits/c1d8638f2ce686d58156437e04988a7fa68e4731

> This created miner builds and a miner release: https://github.com/filecoin-project/lotus/releases/tag/untagged-086fb5cb1c887e1c1658, although the release page is mostly empty except for some titles and the build assets. I cancelled the miner release step before it did its thing so we haven't pushed that: https://github.com/filecoin-project/lotus/actions/runs/9972453588/job/27556032962

This was expected, too, since miner/v1.27.1 hasn't been released before and v1.27.1 is the miner version in the releases branch. The notes for miner releases are sourced from changelog.

> The release page it posted had two titles and two contributor tables; one taken from the CHANGELOG and one generated. The contributors table that was generated was more correct although the title wasn't an h2 (##), I forgot to redo it. But I edited the release page after the draft was created, but when it did the final release it overwrote my edits so I have to edit it again. It would be good if the draft could be used to get it right before it's pushed to the wild since some people get the original via email or other notification where the edits won't be seen after publish.

Removed secondary title in https://github.com/filecoin-project/lotus/pull/12096/commits/55d83f067f9d55421e989e5ced31afa36537c9f7

The release notes are changelog entry + mkreleaselog output. You no longer have to paste the mkreleaselog output in changelog directly.

Added draft notes preservation in https://github.com/filecoin-project/lotus/pull/12096/commits/55d83f067f9d55421e989e5ced31afa36537c9f7

**Comment by rvagg** [comment]
> This was expected, too, since miner/v1.27.1 hasn't been released before and v1.27.1 is the miner version in the releases branch. The notes for miner releases are sourced from changelog.

Oh .. we set the miner version to be something without `-dev`: https://github.com/filecoin-project/lotus/commit/8e7cd7d5d8a1fc79349ffa8d904dc22c803e1912, we probably should leave it as `-dev` so we don't get unwanted releases.

**Comment by BigLep** [comment]
@galargh: are the next steps clear here so we can get this merged?

**Comment by BigLep** [comment]
@rvagg : per slack, @galargh believes this is good to go.  Will you please merge?

**Comment by rvagg** [comment]
Has conflicts that need resolving. @galargh I'm happy for you to squash merge this once you're done with the conflicts.

**Comment by galargh** [comment]
I don't see any outstanding conflicts. Is it with some other branch than the target (master)? Let me know, happy to update as needed.

**Comment by rvagg** [comment]
hm, no, I'm sure there were conflicts when I posted that message! Looks good to go now.

**Comment by galargh** [comment]
Might have been some glitch with GitHub UI; thank you for merging, let me know if there are any issues with it and I'll jump straight on it.

**Pull request merged**
